core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1188,119 +1133,201 @@ class KafkaController(val config: KafkaConfig, zkUtils: ZkUtils, val brokerState
+          val newBrokers = curBrokers.filter(broker => newBrokerIds(broker.id))
+          controllerContext.liveBrokers = curBrokers
+          val newBrokerIdsSorted = newBrokerIds.toSeq.sorted
+          val deadBrokerIdsSorted = deadBrokerIds.toSeq.sorted
+          val liveBrokerIdsSorted = curBrokerIds.toSeq.sorted
+          info("Newly added brokers: %s, deleted brokers: %s, all live brokers: %s"
+            .format(newBrokerIdsSorted.mkString(","), deadBrokerIdsSorted.mkString(","), liveBrokerIdsSorted.mkString(",")))
+          newBrokers.foreach(controllerContext.controllerChannelManager.addBroker)
+          deadBrokerIds.foreach(controllerContext.controllerChannelManager.removeBroker)
+          if(newBrokerIds.nonEmpty)
+            onBrokerStartup(newBrokerIdsSorted)
+          if(deadBrokerIds.nonEmpty)
+            onBrokerFailure(deadBrokerIdsSorted)
+        } catch {
+          case e: Throwable => error("Error while handling broker changes", e)
+        }
+      }
     }
-    partitionsToBeReassigned.foreach { partitionToBeReassigned =>
-      inLock(controllerContext.controllerLock) {
-        if(controller.deleteTopicManager.isTopicQueuedUpForDeletion(partitionToBeReassigned._1.topic)) {
-          error("Skipping reassignment of partition %s for topic %s since it is currently being deleted"
-            .format(partitionToBeReassigned._1, partitionToBeReassigned._1.topic))
-          controller.removePartitionFromReassignedPartitions(partitionToBeReassigned._1)
-        } else {
-          val context = new ReassignedPartitionsContext(partitionToBeReassigned._2)
-          controller.initiateReassignReplicasForTopicPartition(partitionToBeReassigned._1, context)
+  }
+
+  case class TopicChange(topics: Set[String]) extends ControllerEvent {
+    override def process(): Unit = {
+      if (!isActive) return
+      try {
+        val newTopics = topics -- controllerContext.allTopics
+        val deletedTopics = controllerContext.allTopics -- topics
+        controllerContext.allTopics = topics
+
+        val addedPartitionReplicaAssignment = zkUtils.getReplicaAssignmentForTopics(newTopics.toSeq)
+        controllerContext.partitionReplicaAssignment = controllerContext.partitionReplicaAssignment.filter(p =>
+          !deletedTopics.contains(p._1.topic))
+        controllerContext.partitionReplicaAssignment.++=(addedPartitionReplicaAssignment)
+        info("New topics: [%s], deleted topics: [%s], new partition replica assignment [%s]".format(newTopics,
+          deletedTopics, addedPartitionReplicaAssignment))
+        if (newTopics.nonEmpty)
+          onNewTopicCreation(newTopics, addedPartitionReplicaAssignment.keySet)
+      } catch {
+        case e: Throwable => error("Error while handling new topic", e)
+      }
+    }
+  }
+
+  case class PartitionModifications(topic: String) extends ControllerEvent {
+    override def process(): Unit = {
+      if (!isActive) return
```

Comment:
Yup there is a reason. I also wanted to extract this check but there were two issues with doing so:
1. most but not all event types should do the isActive check (think Startup, Elect, Resign).
2. different event types handle the isActive check differently. Most simply return if not active, while others such as the ControlledShutdown prepare a response stating that the controller has moved.

Because of these differences, I couldn't for example naively extract the check out to the ControllerEventThread's doWork. Some options I had considered but ultimately decided against:
- pattern match in the ControllerEventThread's doWork checking event types and conditionally call isActive.
- categorize the event types into those that should have the isActive check and those that shouldn't.  With this categorization, we can let the ControllerEventThread's doWork do the pattern match based on these new types and conditionally run the check.
- categorize the event types into those that should have the isActive check and those that shouldn't.  With this categorization, we can let these new parent types do the check themselves upfront.

I essentially decided against all of these due to the unique behavior of the ControlledShutdown event.